File: full/res/tests/recursive.f
letrec add: Nat -> Nat -> Nat =
  func lhs: Nat (
    if isZero lhs then
      func rhs1: Nat (
        rhs1
      )
    else 
      func rhs2: Nat (
        (add (prd lhs)) (suc rhs2)
      )
  )
in

letrec sum: Nat -> Nat = 
  func i: Nat (
    if isZero i then
      Zero
    else
      (add i) (sum (prd i))
  )
in

letrec fib: Nat -> Nat =
  func o: Nat (
    if isZero o then
      (suc Zero)
    else if isZero (prd o) then
      (suc Zero)
    else 
      (add (fib (prd o))) (fib (prd (prd o)))
  )
in

(fib (suc (suc (suc (suc (suc Zero))))))

// fib (suc (suc (suc (suc (suc Zero)))))

File: full/res/tests/recursive.ml
let rec add: Nat -> Nat -> Nat =
  fun lhs: Nat (
    fun rhs: Nat (
      if isZero lhs then
        rhs
      else
        (add prd lhs) suc rhs
    )
  )
in 

let rec sum: Nat -> Nat = 
  fun n: Nat (
    if isZero n then
      0
    else
      (add n) sum prd n
  )
in 

let rec fib: Nat -> Nat =
  fun n: Nat (
    if isZero n then
      1
    else if isZero prd n then
      1
    else 
      (add fib prd n) fib prd prd n
  )
in
(* This is a comment block.
   (* Nested comments *)
*)

fib (add 4) 5

(* fib (suc (suc (suc (suc (suc Zero))))) *)
